File: backend/app/schemas/substation.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class Topology(BaseModel):
    nodes: list[dict[str, Any]] = Field(default_factory=list)
    edges: list[dict[str, Any]] = Field(default_factory=list)


class SubstationCreate(BaseModel):
    name: str
    sigla: str
    topology: Topology = Field(default_factory=Topology)


class SubstationUpdate(BaseModel):
    user_id: uuid.UUID
    topology: Topology


class SubstationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    sigla: str
    topology: Topology
    version: int
    locked_by: uuid.UUID | None
    locked_by_name: str | None
    locked_at: datetime | None
    created_at: datetime


class SubstationVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    substation_id: uuid.UUID
    topology: Topology
    version: int
    saved_at: datetime
    saved_by: uuid.UUID | None


class LockRequest(BaseModel):
    user_id: uuid.UUID

File: frontend/src/types/substation.ts
export interface TopologyNode {
  id: string;
  type: string;
  position: { x: number; y: number };
  data: Record<string, unknown>;
}

export interface TopologyEdge {
  id: string;
  source: string;
  sourceHandle?: string | null;
  target: string;
  targetHandle?: string | null;
}

export interface Topology {
  nodes: TopologyNode[];
  edges: TopologyEdge[];
}

export interface Substation {
  id: string;
  name: string;
  sigla: string;
  topology: Topology;
  version: number;
  locked_by: string | null;
  locked_by_name: string | null;
  locked_at: string | null;
  created_at: string;
}

export interface SubstationCreate {
  name: string;
  sigla: string;
  topology?: Topology;
}
